### DISCUSSION TITLE
[Detector Support]: Can I use a custom Yolov11 model with CPU detector ?

### DISCUSSION THREAD

**Discussion by lraulet** [blakeblackshear/frigate#14629] [Detector Support] [answered]
### Describe the problem you are having

I trained a Yolov11 model for my specific use case.
Then I used the "yolo export" command to export it in tflite format.

`yolo export model=/home/monhome/yolo/best-v11.pt format=tflite`

However when Frigate starts I get the error : 

`Process detector:cpu1:
Traceback (most recent call last):
  File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/frigate/frigate/edgetpu.py", line 159, in run_detector
    detections = object_detector.detect_raw(input_frame)
  File "/opt/frigate/frigate/edgetpu.py", line 82, in detect_raw
    self.interpreter.set_tensor(self.tensor_input_details[0]["index"], tensor_input)
  File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 572, in set_tensor
    self._interpreter.SetTensor(tensor_index, value)
ValueError: Cannot set tensor: Got value of type UINT8 but expected type FLOAT32 for input 0, name: inputs_0`

### Version

0.11.1-2eada21

### Frigate config file

```yaml
mqtt:
  host: A.B.C.D
  user: myuser
  password: mypassword

cameras:
  camera_cuisine:
    ffmpeg:
      inputs:
        - path: rtsp://E.F.G.H:554/h264Preview_01_main
          roles:
            - detect
            - rtmp            
    detect:
      enabled: True
      width: 640
      height: 640
      fps: 15
    objects:
      track:
         - alone
         - with-prey
    motion:
      mask:
        - 221,26,411,23,414,0,217,0
    rtmp:
      enabled: True
    snapshots:
      enabled: True

    
detectors:
  cpu1:
    type: cpu
model:
  path: /model/best-v11_float32.tflite
  labelmap_path: /model/labelmap.txt
  width: 640
  height: 640
```

### docker-compose file or Docker CLI command

```yaml
services:
  frigate:
    image: "ghcr.io/weltenwort/frigate-synology-dsm7:v0.11.1"
    privileged: true
    hostname: "frigate"
    restart: "unless-stopped"
    shm_size: "128mb"
    devices:
      - "/dev/dri/renderD128" # for intel hwaccel
    volumes:
      - "/dev/bus/usb:/dev/bus/usb"
      - "/etc/localtime:/etc/localtime:ro"
      - "/volume1/docker-frigate/model:/model:ro"

      - "/volume1/docker-frigate/config/config.yml:/config/config.yml:ro"
      - "/volume1/docker-frigate/media:/media/frigate:rw"
      - type: tmpfs # Optional: 1GB of memory, reduces SSD/SD Card wear
        target: /tmp/cache
        tmpfs:
          size: 1000000000
    ports:
      - "8020:5000" # web interface
      - "1936:1935" # RTMP feeds
#    env_file:
#      - stack.env
    environment:
      - LIBVA_DRIVER_NAME=i965 # if hwacl isn't working properly with the default iHD driver
      
#FRIGATE_MQTT_USERNAME=my-frigate-mqtt-username
#FRIGATE_MQTT_PASSWORD=my-frigate-mqtt-password
#FRIGATE_CAMERA_1_RTSP_CREDENTIALS=my-camera-1-username:my-camera-1-password
#FRIGATE_CAMERA_2_RTSP_CREDENTIALS=my-camera-2-username:my-camera-2-password
```

### Relevant Frigate log output

```shell
[2024-10-27 19:40:28] detector.cpu1                  INFO    : Starting detection process: 216
[2024-10-27 19:40:28] frigate.edgetpu                WARNING : CPU detectors are not recommended and should only be used for testing or for trial purposes.
[2024-10-27 19:40:28] frigate.app                    INFO    : Camera processor started for camera_cuisine: 223
[2024-10-27 19:40:28] frigate.app                    INFO    : Capture process started for camera_cuisine: 224
[2024-10-27 19:40:28] ws4py                          INFO    : Using epoll
Process detector:cpu1:
Traceback (most recent call last):
  File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/frigate/frigate/edgetpu.py", line 159, in run_detector
    detections = object_detector.detect_raw(input_frame)
  File "/opt/frigate/frigate/edgetpu.py", line 82, in detect_raw
    self.interpreter.set_tensor(self.tensor_input_details[0]["index"], tensor_input)
  File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 572, in set_tensor
    self._interpreter.SetTensor(tensor_index, value)
ValueError: Cannot set tensor: Got value of type UINT8 but expected type FLOAT32 for input 0, name: inputs_0
```

### Install method

Docker Compose

### Object Detector

CPU (no coral)

### Screenshots of the Frigate UI's System metrics pages

![image](https://github.com/user-attachments/assets/89505b8f-e1ab-40a7-8c42-a37778532c2e)

**Comment by hawkeye217** [comment]
Values are sent as int, but the model was trained on float input. You would either need to make a change to the code or wait for Frigate to support different input configurations.

**Reply by lraulet** [reply]
Sorry, I'm a pure newbie.
Do you mean the model is train on Float (inputs and output) but Frigate read it as Int ?
If yes, why the error seems to say the opposite "waiting for a Float32 but receiving a Uint8" ?

**Reply by hawkeye217** [reply]
The error message can be deceiving because of how models work. But to make it simple, the mismatch is not something that can be easily corrected.

See this comment to understand what would be needed to support INT8 models: https://github.com/ultralytics/ultralytics/issues/5250#issuecomment-1751252920

**Reply by lraulet** [reply]
Ok.
So do you confirm that Frigate (at least for the CPU detector) expect a Uint8 model only ?
If yes, does "SD MobileNet" a good choice ?

**Reply by hawkeye217** [reply, marked as answer]
Correct, Frigate only supports int inputs currently.

**Reply by lraulet** [reply]
Ok. Thx for the confirmation.
I will try moving to "EfficientDet-Lite" model if I find a tuto that explain how I can train it.
Yolo was really easy to do thanks to they CLI.
